[Uniques V2] Feature flags

jsidorenko opened the pull request:
The PR adds a new way to configure the pallet, collection and its items.
- Pallet-level feature flags would allow disabling some functionality that is not needed in the runtime.
- For collections and items I basically extracted fields like is_frozen from the internal structures (ItemDetails, CollectionDetails, etc) and put into the new ItemConfigOf/CollectionConfigOf storage. This would allow extending the functionality of the collections/items without the need to migrate them.
- separated metadata from attributes locking. It was not obvious before that when you lock the metadata that will lock the attributes as well. Now there is an explicit method where you specify what you want to get locked

gilescope commented:
Should we use some other term than feature flags if we're not going to implement them as rust features as that will just get confusing?

jsidorenko commented:
> Should we use some other term than feature flags if we're not going to implement them as rust features as that will just get confusing?

If you check the code, we use `SystemFeatures` for pallet level features and `ItemConfig`/`CollectionConfig` for items and collections

gilescope commented on frame/nfts/src/features/macros.rs:
```diff
@@ -0,0 +1,72 @@
+			}
+		}
+	};
+}
```
This isn't a feature. I'd put this somewhere else. Feels like it should be in primitives somewhere as none of this is NFT related?

jsidorenko replied:
What would be the best file to put this into?

gilescope replied:
Hmm, let's wait till another pallet uses features.

gilescope commented on frame/nfts/src/features/settings.rs:
```diff
@@ -0,0 +1,60 @@
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
```
docs here could give an example of a couple of likely nft settings so that people get the idea.

jsidorenko replied:
done

gilescope replied:
ah so these are part of the feature machinery rather than being a separate settings feature. Wondering if we should one module for features and another for the mechanism.

jsidorenko replied:
@gilescope where would you put them?

gilescope replied:
Here's fine for now. We can revisit locations when we have another pallet that needs the same code.

gilescope commented on frame/nfts/src/functions.rs:
```diff
@@ -36,12 +36,20 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		let collection_details =
 			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
-		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
 		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
```
Would be good to split the `Locked` error into `CollectionLocked` and `ItemLocked` so that UIs can give better feedback to the user.

jsidorenko replied:
Done

gilescope commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1497,6 +1498,10 @@ impl pallet_uniques::Config for Runtime {
 }
 
+parameter_types! {
+	pub FeatureFlags: SystemFeatures = SystemFeatures::empty();
```
if someone asks me what SystemFeatures are I would say pallet features so why don't we rename `SystemFeatures` to `PalletFeatures` ?

jsidorenko replied:
Agree! Renamed ;)

jsidorenko commented on frame/nfts/src/features/lock.rs:
```diff
@@ -0,0 +1,89 @@
+		collection: T::CollectionId,
+		item: T::ItemId,
+		lock_metadata: bool,
+		lock_attributes: bool,
```
You need to have a freezer role to make it non-transferable and call a `freeze_item()` method

jsidorenko replied:
> If we're being consistent with the above then this should take an `ItemConfig` rather than bools?

I made it this way in order to highlight that you can only modify 2 out of 3 settings with that method.
We can change it to ItemConfig and exclude `NonTransferable` if set. I just afraid that could create some confusion.
Wdyt?

jsidorenko commented on frame/nfts/src/features/lock.rs:
```diff
@@ -0,0 +1,89 @@
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
```
done

gilescope commented on frame/nfts/src/features/lock.rs:
```diff
@@ -0,0 +1,92 @@
+		ensure!(origin == details.freezer, Error::<T, I>::NoPermission);
+
+		CollectionConfigOf::<T, I>::try_mutate(collection, |maybe_config| {
+			let config = maybe_config.as_mut().ok_or(Error::<T, I>::UnknownCollection)?;
```
do we need to be able to distinguish this error from the one above? The more specific the error, the less support calls.

jsidorenko replied:
In a lot of places (except the one you pointed out:) I removed the collection existence check and the first validation will be done over the config. So that if I pass a wrong id, I would get the `UnknownCollection` and not a `NoConfig` which would make more sense give the config would always exists together with collection

jsidorenko replied:
Though for that particular place that you highlighted I agree it would make sense to change the error message

gilescope commented on frame/nfts/src/functions.rs:
```diff
@@ -224,6 +240,13 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		let settings = Self::get_item_settings(&collection, &item)?;
+		if settings.is_empty() {
+			ItemConfigOf::<T, I>::remove(&collection, &item);
+		}
```
Is there an attack here where I mint and burn lots of items getting my deposit back but I rack up settings in the state?

ggwpez commented on frame/nfts/src/features/freeze.rs:
```diff
@@ -0,0 +1,62 @@
+/// It's also possible to revert the setting back.
+/// An origin must have a `Freezer` role in order to call those methods.
+impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	pub fn do_freeze_item(
```
The calls should be the only public functions on a pallet AFAIK. Everything else `pub(crate)`.

jsidorenko replied:
What if I want to call some function from another pallet? Afaik, the main usage of the uniques pallet is to provide the basic implementation to deal with NFTs and all the custom logic will be placed in other pallets.

gilescope replied:
Any other pallet building on it should be able to go through the public api.

ggwpez replied:
Through traits

jsidorenko replied:
Done

ggwpez commented on frame/nfts/src/functions.rs:
```diff
@@ -138,6 +145,9 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			CollectionMaxSupply::<T, I>::remove(&collection);
+			CollectionConfigOf::<T, I>::remove(&collection);
+			#[allow(deprecated)]
+			ItemConfigOf::<T, I>::remove_prefix(&collection, None);
```
The PoV Size of this is potentially unlimited. This is basically the same problem as https://github.com/paritytech/substrate/pull/12310 which favours a multi-step approach.

gilescope replied:
Maybe one of the standard Error enums should be `Error::IncompletePleaseCallAgain` ?

ggwpez replied:
The witness should probably contain an upper limit to bound the weight.

ggwpez commented on frame/nfts/src/impl_nonfungibles.rs:
```diff
@@ -78,26 +78,40 @@ impl<T: Config<I>, I: 'static> Inspect<<T as SystemConfig>::AccountId> for Palle
 	) -> DispatchResult {
+		let mut settings = *settings;
+		// FreeHolding could be set by calling the force_create() only
+		if settings.contains(CollectionSetting::FreeHolding) {
```
I think it would be better to have a wrapper type for clearer semantics:  
```rust
settings.is_enabled(…)
settings.is_disabled(…)
```

ggwpez commented on frame/nfts/src/lib.rs:
```diff
@@ -333,6 +314,23 @@ pub mod pallet {
 		OptionQuery,
 	>;
 
+	#[pallet::storage]
+	/// Config of a collection.
```
```suggestion
	/// Config of a collection.
	#[pallet::storage]
```

ggwpez commented on frame/nfts/src/lib.rs:
```diff
@@ -333,6 +314,23 @@ pub mod pallet {
+	pub(super) type CollectionConfigOf<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Blake2_128Concat, T::CollectionId, CollectionConfig, OptionQuery>;
+
+	#[pallet::storage]
+	/// Config of an item.
```
```suggestion
	/// Config of an item.
	#[pallet::storage]
```

ggwpez commented on frame/nfts/src/types.rs:
```diff
@@ -40,6 +43,12 @@ pub(super) type ItemTipOf<T, I = ()> = ItemTip<
 >;
 
+pub trait Incrementable {
+	fn increment(&self) -> Self;
```
```suggestion
	fn saturating_increment(&self) -> Self;
```

jsidorenko replied:
Let's do that during the final PR to master

ggwpez commented on frame/nfts/src/types.rs:
```diff
@@ -40,6 +43,12 @@ pub(super) type ItemTipOf<T, I = ()> = ItemTip<
+	fn increment(&self) -> Self;
+	fn initial_value() -> Self;
+}
+impl_incrementable!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
```
You can probably implement this for all `Saturating + Zero` types.

jsidorenko replied:
same as above

ggwpez commented on frame/nfts/src/types.rs:
```diff
@@ -172,3 +171,89 @@ pub struct PriceWithDirection<Amount> {
 }
+
+// Support for up to 64 user-enabled features on a collection.
+#[bitflags]
```
I dont trust this macro. Could you add some test that checks that the values are indeed exclusive? So that they have no overlapping bits?

gilescope replied:
this is a good point, we should axe enumflags2 crate in favour of bitflags crate as that's more commonly used in substrate. less deps is better security.

jsidorenko replied:
We already use that macro in the identity pallet, it's not new here

ggwpez commented on frame/nfts/src/types.rs:
```diff
@@ -172,3 +171,89 @@ pub struct PriceWithDirection<Amount> {
+}
+impl_codec_bitflags!(CollectionConfig, u64, CollectionSetting);
+
+// Support for up to 64 user-enabled features on an item.
```
```suggestion
/// Support for up to 64 user-enabled features on an item.
```

ggwpez commented on frame/support/src/traits/tokens/nonfungible_v2.rs:
```diff
@@ -0,0 +1,204 @@
+use sp_std::prelude::*;
+
+/// Trait for providing an interface to a read-only NFT-like set of items.
+pub trait Inspect<AccountId> {
```
How can I check if an item exists?

ggwpez commented on frame/support/src/traits/tokens/nonfungible_v2.rs:
```diff
@@ -0,0 +1,204 @@
+>(sp_std::marker::PhantomData<(F, A, AccountId)>);
+
+impl<
+		F: nonfungibles::Inspect<AccountId>,
```
Writing it with a `where` clause makes it a bit more readable IHMO.

ggwpez commented on frame/support/src/traits/tokens/nonfungible_v2.rs:
```diff
@@ -0,0 +1,204 @@
+		<F as nonfungibles::Inspect<AccountId>>::attribute(&A::get(), item, key)
+	}
+	fn typed_attribute<K: Encode, V: Decode>(item: &Self::ItemId, key: &K) -> Option<V> {
+		<F as nonfungibles::Inspect<AccountId>>::typed_attribute(&A::get(), item, key)
```
This only works if the collection ID is known at compile time, or?

ggwpez commented on frame/support/src/traits/tokens/nonfungibles_v2.rs:
```diff
@@ -0,0 +1,243 @@
+}
+
+/// Trait for providing the ability to destroy collections of nonfungible items.
+pub trait Destroy<AccountId>: Inspect<AccountId> {
```
Why does a `Destroy` need to be an `Inspect`?

jsidorenko replied:
Idk, it's a copy of the existing `frame/support/src/traits/tokens/nonfungibles.rs` I only added support for feature flags

ggwpez commented on frame/support/src/traits/tokens/nonfungibles_v2.rs:
```diff
@@ -0,0 +1,243 @@
+	/// Attempt to set the strongly-typed attribute `value` of `collection`'s `key`.
+	///
+	/// By default this just attempts to use `set_attribute`.
+	fn set_typed_collection_attribute<K: Encode, V: Encode>(
```
These traits probably need more a in-depth review than I have time to do rn.

jsidorenko replied:
Since I only copy+pasted them from v1, let's work on them in a separate PR, wdyt?

ggwpez reviewed (approved):
Look good from afar, but should be fine since it merges into an internal branch.

jsidorenko commented on frame/nfts/src/features/atomic_swap.rs:
```diff
@@ -31,6 +31,10 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		duration: <T as SystemConfig>::BlockNumber,
 	) -> DispatchResult {
+		ensure!(
+			Self::is_pallet_feature_disabled(PalletFeature::NoSwaps),
```
It's because by default the flags are empty and we only store what is disabled. This applies to all the settings: pallet's, collection's, item's.

ggwpez replied:
Why? Does not matter whether you use `0...` or `1...` as default.

jsidorenko replied:
done

jsidorenko commented on frame/nfts/src/features/lock.rs:
```diff
@@ -0,0 +1,92 @@
+		collection: T::CollectionId,
+		lock_config: CollectionConfig,
+	) -> DispatchResult {
+		let details =
```
Yep, this allows to change different collection-level settings, I assume that method should always be accessible

gilescope replied:
we could state that in the docs here so people know why it should not be turned into a feature?

gilescope commented on frame/nfts/src/lib.rs:
```diff
@@ -499,8 +499,10 @@ pub mod pallet {
 		WrongOwner,
 		/// Invalid witness data given.
 		BadWitness,
```
```suggestion
		/// The witness data given does not match the current state of the chain.
		BadWitness,
```

jsidorenko replied:
Done

gilescope commented on frame/nfts/src/lib.rs:
```diff
@@ -499,8 +499,10 @@ pub mod pallet {
+		CollectionIdInUse,
+		/// Items within that collection are non-transferable.
+		ItemsNotTransferable,
 		/// The item or collection is frozen.
```
would be good to say what 'frozen' means here. Be good to state the difference between frozen and locked.

jsidorenko replied:
Done

gilescope commented on frame/nfts/src/lib.rs:
```diff
@@ -1262,20 +1272,35 @@ pub mod pallet {
 			value: BoundedVec<u8, T::ValueLimit>,
 		) -> DispatchResult {
+			ensure!(
+				Self::is_pallet_feature_disabled(PalletFeature::NoAttributes),
```
Shouldn't features be always positive like rust features? I.e. the feature should be Attributes - it could be a default feature.

gilescope replied:
This is the same thing as Oliver's point above.

gilescope commented on frame/nfts/src/tests.rs:
```diff
@@ -398,53 +426,57 @@ fn set_item_metadata_should_work() {
+			Error::<Test>::UnknownCollection,
 		);
-		assert_ok!(Nfts::clear_metadata(RuntimeOrigin::signed(1), 0, 42));
+		assert_ok!(Nfts::clear_metadata(RuntimeOrigin::root(), 0, 42));
```
Why does this now have to be root?

jsidorenko replied:
When the metadata is locked, only the root has rights to clear it. There is a check for that a couple of line above:
```
		assert_noop!(
			Nfts::clear_metadata(RuntimeOrigin::signed(1), 0, 42),
			Error::<Test>::LockedItemMetadata,
		);
```

jsidorenko commented on frame/nfts/src/features/macros.rs:
```diff
@@ -0,0 +1,72 @@
+		$(
+			impl Incrementable for $type {
+				fn increment(&self) -> Self {
+					self.saturating_add(1)
```
done

jsidorenko commented on frame/nfts/src/features/freeze.rs:
```diff
@@ -0,0 +1,62 @@
+use crate::*;
+use frame_support::pallet_prelude::*;
+
+/// Freeze functions allow to make particular items non-transferable.
```
done

jsidorenko commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1497,6 +1498,10 @@ impl pallet_uniques::Config for Runtime {
 }
 
+parameter_types! {
+	pub Features: PalletFeatures = PalletFeatures::empty();
```
done

jsidorenko commented on frame/nfts/src/lib.rs:
```diff
@@ -512,7 +514,15 @@ pub mod pallet {
 		/// The named owner has not signed ownership of the collection is acceptable.
```
`/// The named owner has not signed ownership acceptance of the collection.` Is that better? :)

jsidorenko commented:
@gilescope @ggwpez Guys, I re-wrote the whole implementation, swapped negative setting names with positive and added different helpers to make the code more readable. Plus I renamed freeze/lock/thaw methods to remove the ambiguity.
Could you pls review this PR again?

gilescope reviewed (approved):
Excellent, most happy.

ggwpez commented on frame/nfts/src/functions.rs:
```diff
@@ -36,12 +36,22 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			Collection::<T, I>::get(&collection).ok_or(Error::<T, I>::UnknownCollection)?;
-		ensure!(!collection_details.is_frozen, Error::<T, I>::Frozen);
-		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::Locked);
+		ensure!(!T::Locker::is_locked(collection, item), Error::<T, I>::ItemLocked);
```
The `Frozen` error enum constant is not used anymore.

ggwpez commented on frame/nfts/src/lib.rs:
```diff
@@ -565,19 +576,29 @@ pub mod pallet {
 
+			let mut config = config;
+			// DepositRequired could be disabled by calling the force_create() only
+			if config.has_disabled_setting(CollectionSetting::DepositRequired) {
+				config.enable_setting(CollectionSetting::DepositRequired);
+			}
```
```suggestion
config.enable_setting(CollectionSetting::DepositRequired);
```
I think it would be better to error here though. Otherwise the creator will wonder why it is enabled.

jsidorenko replied:
Nice suggestion!

ggwpez commented on frame/nfts/src/types.rs:
```diff
@@ -172,3 +176,128 @@ pub struct PriceWithDirection<Amount> {
+
+impl CollectionConfig {
+	pub fn all_settings_enabled() -> Self {
+		Self(BitFlags::EMPTY)
```
Normally bitflags are used in a way that a 1 signals enabled and 0 disabled.  
If you use `EMPTY` as enabled, then this will require a storage migration if you add a new setting, unless you implicitly want to force-enable it onto all deployed nfts+collections.  
1 = enabled is more defensive.

jsidorenko replied:
You set 1 (true) for the disabled things, so if nothing is disabled, then there is no need to store anything.
Why would you need to run a migration for the new settings? If you add smth new, it will be enabled by default, until you explicitly disable it

ggwpez commented on frame/nfts/src/benchmarking.rs:
```diff
@@ -146,25 +145,19 @@ benchmarks_instance_pallet! {
 
 	destroy {
 		let n in 0 .. 1_000;
-		let m in 0 .. 1_000;
```
In the weight file it looks like these components are used. Why did you remove them?

jsidorenko replied:
Because previously it was possible to create metadata and attributes for non-existing records and that benchmark worked. After I added a check whether the record exists, that benchmark started to fail when we try to add metadata (or an attribute) in separate loop

jsidorenko commented on frame/nfts/src/features/lock.rs:
```diff
@@ -0,0 +1,120 @@
+		CollectionConfigOf::<T, I>::try_mutate(collection, |maybe_config| {
+			let config = maybe_config.as_mut().ok_or(Error::<T, I>::NoConfig)?;
+
+			if lock_config.has_disabled_setting(CollectionSetting::TransferableItems) {
+				config.disable_setting(CollectionSetting::TransferableItems);
+			}
+			if lock_config.has_disabled_setting(CollectionSetting::UnlockedMetadata) {
+				config.disable_setting(CollectionSetting::UnlockedMetadata);
+			}
+			if lock_config.has_disabled_setting(CollectionSetting::UnlockedAttributes) {
+				config.disable_setting(CollectionSetting::UnlockedAttributes);
+			}
```
I'm checking only those 3 settings since the last one (`DepositRequired`) can be set by the root only. In the future, I expect more settings with custom business logic or gated control

jsidorenko replied:
> With bit flags something config &= !lock_config should work. Or more Rusty and less Cy: config -= lock_config.

Wouldn't that significantly reduce the code readability? Giles wrote above that we should write a more readable code, that's why I added `has_disabled_setting()`, `disable_setting()` and other helpers
